networkx/linalg/algebraicconnectivity.py

```diff
@@ -244,7 +244,7 @@ def P(x):
             W -= (W.T * X * X.T).T
             project(W)
             # Compute the diagonal of P * L * P as a Jacobi preconditioner.
-            D = L.diagonal()
+            D = L.diagonal().astype(float)
             D += 2. * (asarray(X) * asarray(W)).sum(axis=1)
```

Comment:
Reinitializing the solver is as simple as

```python
solver = _PCGSolver(lambda x: P(L * P(x)), lambda x: D * x)
```

You just create a new solver. For PCG, this is cheap enough that we can afford to do that many times.

As for `astype(float)` vs `+=`, now that `astype(float)` has been merged, I think that we should just keep it unless there is a compelling reason use `+=`.